### PR TITLE
Update Python code to use Python 3.11

### PR DESCRIPTION
**Pull request by dhruvkb**
## Fixes

Related to #663
Related to #1370

## Description

This PR updates all packages except the catalog to use Python 3.11. The catalog is being held back by Airflow, which does not support 3.11 yet.

## Testing Instructions

CI should pass.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added or updated tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [ ] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Review by krysal — COMMENTED** [review]
It's looking good! I assume this will need to be rebased after #1225 is merged.

**Comment by dhruvkb** [comment]
@krysal rebased. The `Dockerfile`s for the API and ingestion server become unchanged after updating because the ones on `main` can automatically use the new version from the `Pipfile`s.

File: .github/workflows/label_pr.yml [hunk 1]
```diff
@@ -20,17 +20,11 @@ jobs:
...
-          python-version: "3.10"
-
-      - name: Install dependencies
-        working-directory: ./automations/python
-        run: |
-          python -m pip install --user --upgrade pip
-          python -m pip install --user pipenv
-          pipenv install --deploy
+          setup_nodejs: false # Node.js is not needed to run Python automations.
+          install_recipe: "automations/python/install"
```
Comment (krysal) [inline .github/workflows/label_pr.yml:27]: Great simplification here!

File: .github/workflows/label_pr.yml [hunk 2]
```diff
@@ -20,17 +20,11 @@ jobs:
       - name: Checkout repository
         uses: actions/checkout@v3
 
-      - name: Setup Python
-        uses: actions/setup-python@v4
+      - name: Setup CI env
```
Comment (krysal) [inline .github/workflows/label_pr.yml:23]:
```suggestion
      - name: Setup CI environment
```

File: documentation/Pipfile [hunk 3]
```diff
@@ -4,7 +4,6 @@ verify_ssl = true
 name = "pypi"
 
 [packages]
-sphinxcontrib-mermaid = "*"
 
 [dev-packages]
 sphinx = "*"
```
Comment (krysal) [inline documentation/Pipfile:9]: Curious that all the documentation packages are in the dev section.

**Review by krysal — APPROVED** [review]
Awesome! LGTM.

**Review by obulat — APPROVED** [review]
LGTM!

File: documentation/Pipfile (hunk 3 above)
Comment (dhruvkb) [reply documentation/Pipfile:9]: The build output are static pages so everything is a dev dependency.

File: .github/workflows/label_pr.yml (hunk 2 above)
Comment (dhruvkb) [reply .github/workflows/label_pr.yml:23]: We're using the same name everywhere so it's more consistent this way.

**Pull request merged**